sycl/include/sycl/builtins.hpp

```diff
@@ -789,6 +896,79 @@ detail::enable_if_t<detail::is_svgenfloat<T>::value, T> sign(T x) __NOEXC {
   return __sycl_std::__invoke_sign<T>(x);
 }
 
+// marray common functions
+
+// TODO: can be optimized in the way math functions are optimized (usage of
+// vec<T, 2>)
+#define __SYCL_MARRAY_COMMON_FUNCTION_UNOP_OVERLOAD(NAME, ARGTYPE, ARGVAL,     \
+                                                    ITEM)                      \
+  template <typename T,                                                        \
+            typename = std::enable_if_t<detail::is_mgenfloat<T>::value>>       \
+  sycl::marray<detail::marray_element_type<T>, T::size()> NAME(ARGTYPE ARGVAL) \
+      __NOEXC {                                                                \
+    sycl::marray<detail::marray_element_type<T>, T::size()> res;               \
+    for (int i = 0; i < ARGVAL.size(); i++) {                                  \
+      res[i] = NAME(ITEM);                                                     \
+    }                                                                          \
+    return res;                                                                \
+  }
+
+#define __SYCL_MARRAY_COMMON_FUNCTION_BINOP_OVERLOAD(NAME, ARG1, ARG2, ITEM1,  \
+                                                     ITEM2)                    \
+  template <typename T,                                                        \
+            typename = std::enable_if_t<detail::is_mgenfloat<T>::value>>       \
+  sycl::marray<detail::marray_element_type<T>, T::size()> NAME(ARG1, ARG2)     \
+      __NOEXC {                                                                \
+    sycl::marray<detail::marray_element_type<T>, T::size()> res;               \
+    for (int i = 0; i < x.size(); i++) {                                       \
```

Comment:
`x.size()` is unreliable: it assumes that one of `ARGN` arguments will be named `x`. We should either use `res.size()` or `T::size()` here.

Reply:
886dd84